installation-manager/src/main/java/org/wildfly/core/instmgr/InstMgrResourceDefinition.java

```diff
@@ -0,0 +1,378 @@
+                .setCapabilities(INSTALLATION_MANAGER_CAPABILITY).setRuntime());
+        this.imf = imf;
+        this.imService = imService;
+        this.readHandler = new ReadHandler(imService);
+        this.writeHandler = new WriteHandler(imService);
+    }
+
+    @Override
+    public void registerOperations(ManagementResourceRegistration resourceRegistration) {
+        super.registerOperations(resourceRegistration);
+
+        InstMgrHistoryHandler historyHandler = new InstMgrHistoryHandler(imService, imf);
+        resourceRegistration.registerOperationHandler(InstMgrHistoryHandler.DEFINITION, historyHandler);
+
+        InstMgrCreateSnapshotHandler createSnapshotHandler = new InstMgrCreateSnapshotHandler(imService, imf);
+        resourceRegistration.registerOperationHandler(InstMgrCreateSnapshotHandler.DEFINITION, createSnapshotHandler);
+
+        InstMgrListUpdatesHandler lstUpdatesHandler = new InstMgrListUpdatesHandler(imService, imf);
+        resourceRegistration.registerOperationHandler(InstMgrListUpdatesHandler.DEFINITION, lstUpdatesHandler);
+
+        InstMgrCleanHandler clean = new InstMgrCleanHandler(imService, imf);
+        resourceRegistration.registerOperationHandler(InstMgrCleanHandler.DEFINITION, clean);
+
+        InstMgrPrepareUpdateHandler prepUpdatesHandler = new InstMgrPrepareUpdateHandler(imService, imf);
+        resourceRegistration.registerOperationHandler(InstMgrPrepareUpdateHandler.DEFINITION, prepUpdatesHandler);
+
+        InstMgrPrepareRevertHandler revertHandler = new InstMgrPrepareRevertHandler(imService, imf);
+        resourceRegistration.registerOperationHandler(InstMgrPrepareRevertHandler.DEFINITION, revertHandler);
+
+        InstMgrRemoveChannelHandler channelRemoveHandler = new InstMgrRemoveChannelHandler(imService, imf);
+        resourceRegistration.registerOperationHandler(channelRemoveHandler.DEFINITION, channelRemoveHandler);
+
+        InstMgrCustomPatchHandler customPatchHandler = new InstMgrCustomPatchHandler(imService, imf);
+        resourceRegistration.registerOperationHandler(customPatchHandler.DEFINITION, customPatchHandler);
+    }
+
+    @Override
+    public void registerAttributes(ManagementResourceRegistration resourceRegistration) {
+        resourceRegistration.registerReadWriteAttribute(CHANNELS, readHandler, writeHandler);
+    }
+
+    private class WriteHandler implements OperationStepHandler {
+        private final InstMgrService imService;
+
+        public WriteHandler(InstMgrService imService) {
+            this.imService = imService;
+        }
+
+        @Override
+        public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+            final List<ModelNode> channelsListMn = new ArrayList<>();
+            ModelNode operationName = operation.get(OP);
+            if (!operationName.asString().equals(UNDEFINE_ATTRIBUTE_OPERATION)) {
```

Comment:
The undefine-attribute op is just a shorthand for the perfectly valid write-attribute with an undefined value; we can't assume the shorthand is used. You need to test VALUE, e.g.

if (operation.hasDefined(VALUE))